benchmarks/dataset_serialize_benchmark.py

```diff
@@ -0,0 +1,272 @@
+
+        cases = self.get_cases(case, kwargs)
+
+        for source in self.get_sources(source):
+
+            log.info("source %s: download, if required", source.name)
+            source.download_source_if_not_exists()
+            tags = self.get_tags(kwargs, source)
+
+            t0 = time.monotonic()
+            source_ds = self._get_dataset_for_source(source)
+            log.info(
+                "constructed Dataset object for source in %.4f s", time.monotonic() - t0
+            )
+
+            for case in cases:
+
+                log.info("case %s: create directory", case)
+                dirpath = self._create_tmpdir_in_ramdisk(case)
+                log.info("directory created, path: %s", dirpath)
+
+                yield self.benchmark(
+                    f=self._get_benchmark_function(
+                        case, source.name, source_ds, dirpath
+                    ),
+                    extra_tags=tags,
+                    options=kwargs,
+                    case=case,
+                )
+
+                # Free up memory in the RAM disk (tmpfs), assuming that we're
+                # otherwise getting close to filling it (depending on the
+                # machine this is executed on, a single test might easily
+                # occupy 10 % or more of this tmpfs). Note that what
+                # accumulated in `dirpath` is the result of potentially
+                # multiple iterations.
+                self._report_dirsize_and_wipe(dirpath)
+
+        # Finally, remove outest directory. Should have no contents by now, but
+        # if an individual benchmark iteration was Ctrl+C'd then this here
+        # might still do useful cleanup.
+        self._report_dirsize_and_wipe(OUTPUT_DIR_PREFIX)
+
+    def _get_benchmark_function(
+        self, case, source_name: str, source_ds: ds.Dataset, dirpath: str
+    ):
+
+        (selectivity, serialization_format) = case
+
+        # Option A: read-from-disk -> deserialize -> filter -> into memory
+        # before timing serialize -> write-to-tmpfs
+        t0 = time.monotonic()
+        data = source_ds.to_table(filter=self.filters[source_name][selectivity])
+        log.info("read source dataset into memory in %.4f s", time.monotonic() - t0)
+
+        # Option B (thrown away, but kept for posterity): use a Scanner()
+        # object to transparently filter the source dataset upon consumption,
+        # in which case what's timed is read-from-disk -> deserialize -> filter
+        # -> serialize -> write-to-tmpfs
+        #
+        # Note(JP): I have confirmed that for the data used in this benchmark
+        # this option is dominated by read-from-disk to the extent that no
+        # useful signal is generated for the write phase, at least for some
+        # serialization formats.
+        #
+        # data = pyarrow.dataset.Scanner.from_dataset(source_ds,
+        #     filter=self.filters[source_name][selectivity])
```

Comment:
I kept this because it's somewhat non-obvious from pyarrow docs that this `Scanner()`-based technique really works (that the `Scanner` instance can be provided as `data` argument to `write_dataset()`). I tried it out and it looks good, it's certainly a lazy evaluation. It's a super cool technique that maybe we should blog about (well, maybe my enthusiasm is too big because I am new to this ecosystem :)).

I think I'd like to add a benchmark that does the complete round trip in a child process and then the metric we would care about is the maximum memory usage of that child process, confirming that this really is operating in a stream-like fashion: that at any given time the process only holds a fraction of the complete data set in memory.